src/layout/actions/layout.rs

```diff
@@ -880,15 +882,24 @@ impl LayoutTree {
     /// Draws the borders recursively, down from the top to the bottom.
     fn draw_borders_rec(&mut self, mut children: Vec<NodeIndex>)
                         -> CommandResult {
+        let mut updated_nodes: HashSet<NodeIndex> = HashSet::from_iter(children.iter().cloned());
+
         while children.len() > 0 {
             let child_ix = children.pop().unwrap();
-            children.extend(self.tree.grounded_children(child_ix));
+
+            for child in self.tree.grounded_children(child_ix) {
+                if !updated_nodes.contains(&child) {
+                    updated_nodes.insert(child);
+                    children.push(child);
+                }
+            }
+
             let parent_ix = self.tree.parent_of(child_ix)
                 .expect("Node had no parent");
             let children = self.tree.children_of(parent_ix);
             let index = children.iter().position(|&node_ix| node_ix == child_ix)
                 .map(|num| (num + 1).to_string());
-            if Some(child_ix) != self.active_container {
+            if !self.tree.on_path(child_ix) {
```

Comment:
Woops, spoke to soon. Found one regression with your changes. You can't use the `on_path` check, you have to check the `active_container` directly, otherwise it breaks border focus when a subsurface is spawned for a view.

A subsurface is usually a drop down menu, or a right click menu, or one of those popup "hover" bubbles. They are views in their own right, but they have another view as their parent. While they are treated as full views (e.g they have a path in the layout tree), they don't update the `active_container`. This is a bit of a hack, but the active path will _not_ always point to the `active_container` (yes that's confusing, I should perhaps re-think that later).

So if you could revert this change (to do the original `Some(child_ix) != self.active_container` check) that would fix this issue.

Reply:
Ok. I'll take a look

Reply:
About the change with vector drain: Don't you think it will be slow? Having just a hashset of indices is one thing. Creating a set every time and then vector from it may be too slow. Because of memory fragmentation for example. I don't like this condition with 2 pushes eather, but I'd say it's better. 

Maybe there is a reason to use some function to redraw just visible containers, not to pass children? It's a tree, so a list of visible containers will be relatively small. 
And the function will be much simpler. Something like:
```rust
while !visible.empty() {
    let node = visible.pop();
    if node.type == Stacked {
        let index = container.iter().position(|&node| node.ix == visible.head()));
        b.set_title(format!("{:?} ({}/{})", layout, index.unwrap_or("?".into()), node.children.len())
    }
}
```

What do you think about this?

Reply:
But as I understand, it doesn't fix an issue with my change. Maybe to store a list of visible containers instead of active container? Use a list of active containers? )

Reply:
What do you think about another condition
```rust
Some(child_ix) != self.active_container || !self.tree.on_path(child_ix) {
    self.set_borders(child_ix, borders::Mode::Inactive)?;
}
````
So if it's not on path or not and active container (container with popup will remain active), set inactive.

Reply:
A function to return visible children could be helpful and probably improve performance. There would need to be more testing necessary by me to ensure that there aren't any other border focusing bugs (there are weird edge cases, one of which I can think of is switching focus from one window to a window in another output. The first window needs to be "inactive" now, so it needs to still be considered visible even though its not in the active output).

If you want to make this change, I suggest you open a different PR. Lets keep these changes focused on this PR, and if you want to add the "visible" list we can talk about it on gitter to see what other tiling methods would benefit from this change.